src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -284,10 +293,20 @@ private static boolean propagateContextIfNotPresent(ServiceInvocationContext pro
             return false;
         }, () -> {
             setCurrent(propagatedContext);
+            if (propagatedContext.makeContextAwareHandlers != null) {
+                propagatedContext.makeContextAwareHandlers.forEach(MakeContextAwareHandler::enter);
+            }
             return true;
         });
     }
 
+    public void addMakeContextAwareHandler(Runnable onContextEnter, Runnable onContextExit) {
+        if (makeContextAwareHandlers == null) {
+            makeContextAwareHandlers = new ArrayList<>();
+        }
+        makeContextAwareHandlers.add(new MakeContextAwareHandler(onContextEnter, onContextExit));
```

Comment:
Instead of creating a new wrapper object for each `add...Handler()`, could we keep two lists - one for enter and the other for exit?